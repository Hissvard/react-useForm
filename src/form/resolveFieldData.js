import { Map } from 'immutable'
import { fieldsKey } from './reducer/fieldReducer'
import {
  useBooleanField,
  useNumberField,
  useListField,
  useSelectField,
  useTextField,
  useObjectField,
} from './fields'

export const resolveFieldData = (state, dispatch) => {
  const fieldData = state.get(fieldsKey, Map()).entrySeq().reduce((acc, [fieldName, current]) => {
    const fieldState = current.set('getAllValues', () => getFieldValues(fieldData))

    acc[fieldName] = resolveField(fieldState, dispatch)
    return acc
  }, {})
  return fieldData
}

export const resolveField = (fieldState, dispatch) => {
  const field = fieldState.getIn(['initial', 'field'])
  const fieldType = fieldState.getIn(['initial', 'type'])
  switch (fieldType) {
    case 'select':
      return useSelectField(fieldState, dispatch, field)
    case 'boolean':
      return useBooleanField(fieldState, dispatch, field)
    case 'number':
      return useNumberField(fieldState, dispatch, field)
    case 'list':
      return useListField(fieldState, dispatch, field)
    case 'object':
      return useObjectField(fieldState, dispatch, field)
    case 'text':
    default:
      return useTextField(fieldState, dispatch, field)
  }
}

export const getFieldValues = (fieldData) => {
  return Object.entries(fieldData).reduce((acc, [k, v]) => {
    acc[k] = v.props.value
    return acc
  }, {})
}

export const getFieldProps = (fieldData, state = Map()) => {
  return Object.entries(fieldData).reduce((acc, [key, v]) => {
    const passThrough = state.getIn([fieldsKey, key, 'initial', 'field'], {}).passThrough || {}

    acc[key] = {
      ...v.props,
      ...passThrough,
    }
    return acc
  }, {})
}
