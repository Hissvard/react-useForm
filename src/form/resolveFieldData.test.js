import { Map } from 'immutable'
import { getFieldProps } from './resolveFieldData'

let fieldData
beforeEach(() => {
  fieldData = {
    one: { props: { value: 'one', onChange: 'onChange-one' } },
    two: { props: { value: 'two', onChange: 'onChange-two' } },
  }
})

describe('getFieldProps', () => {

  test('reduces all props on keys', () => {
    expect(getFieldProps(fieldData)).toMatchSnapshot()
  })

  test('spreads passThrough from state if it exists', () => {
    fieldData = {
      one: { props: { value: 'one', onChange: 'onChange-one' } },
      two: { props: { value: 'two', onChange: 'onChange-two' } },
    }

    const state = Map([
      ['fields', Map([
        ['one', Map([
          ['initial', Map([
            ['field', { passThrough: { key: 'key-one' } }],
          ])],
        ])],
      ])],
    ])
    expect(getFieldProps(fieldData, state)).toMatchSnapshot()
  })
})
