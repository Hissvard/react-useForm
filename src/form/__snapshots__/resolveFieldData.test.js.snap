// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFieldProps reduces all props on keys 1`] = `
Object {
  "one": Object {
    "onChange": "onChange-one",
    "value": "one",
  },
  "two": Object {
    "onChange": "onChange-two",
    "value": "two",
  },
}
`;

exports[`getFieldProps spreads passThrough from state if it exists 1`] = `
Object {
  "one": Object {
    "key": "key-one",
    "onChange": "onChange-one",
    "value": "one",
  },
  "two": Object {
    "onChange": "onChange-two",
    "value": "two",
  },
}
`;
